system/emergency_handler/src/emergency_handler/emergency_handler_core.cpp

```diff
@@ -151,14 +178,104 @@ void EmergencyHandler::publishControlCommands()
     }
     pub_gear_cmd_->publish(msg);
   }
+}
 
-  // Publish Emergency State
-  {
-    autoware_auto_system_msgs::msg::EmergencyState emergency_state;
-    emergency_state.stamp = stamp;
-    emergency_state.state = emergency_state_;
-    pub_emergency_state_->publish(emergency_state);
+void EmergencyHandler::publishMRMState()
+{
+  mrm_state_.stamp = this->now();
+  pub_mrm_state_->publish(mrm_state_);
+}
+
+void EmergencyHandler::operateMRM()
+{
+  using autoware_adapi_v1_msgs::msg::MRMState;
+
+  if (mrm_state_.state == MRMState::NORMAL) {
+    // Cancel MRM behavior when returning to NORMAL state
+    const auto current_mrm_behavior = MRMState::NONE;
+    if (current_mrm_behavior != mrm_state_.behavior) {
+      cancelMRMBehavior(mrm_state_.behavior);
+      mrm_state_.behavior = current_mrm_behavior;
+    }
+    return;
+  }
+  if (mrm_state_.state == MRMState::MRM_OPERATING) {
+    const auto current_mrm_behavior = getCurrentMRMBehavior();
+    if (current_mrm_behavior != mrm_state_.behavior) {
+      cancelMRMBehavior(mrm_state_.behavior);
+      callMRMBehavior(current_mrm_behavior);
```

Comment:
[IMO] I personally feel that instead of call/cancel MRM via services, we should receive MRM commands continuously via topics and switch them depending on the emergency status. I'm worried about the latency here.

But if you believe this is the right way, it's okay. I'll leave it to you.

Reply:
![image](https://user-images.githubusercontent.com/876355/200782334-4db4c67f-1f7f-4def-9ef1-ca9cd8ad8b06.png)

I measured the time between the system error monitor's determination of an emergency and the output of an emergency stop command (n=1).
The delay due to services is about 20msec, which is small compared to the delay caused by timer callback on other nodes, so I will proceed with this implementation.

Reply:
Okay, thank you for the explanation!